Tests  setup jsdom testing

mcksheridan opened the pull request:
## Description
(This is just a draft, so please don't merge it.)
I started working on [FSA2021-69](https://sparkbox.atlassian.net/browse/FSA2021-69), but since the tests for [FSA2021-68](https://sparkbox.atlassian.net/browse/FSA2021-68) hadn't been added yet, I circled back to add those instead.
I used the [jsdom](https://github.com/jsdom/jsdom) to perform tests that manipulated the DOM API. I ultimately added four tests, one to layout a basic implementation of jsdom, and the other three to test the `handleClickCell()` function.

## Spec

See Issue: [FSA2021-68](https://sparkbox.atlassian.net/browse/FSA2021-68)

## Validation
<!-- delete anything irrelevant to this PR -->

- ✔️  This PR has code changes, and our linters still pass.
- ✔️  This PR has new code, so new tests were added or updated, and they pass.

### To Validate

1. Make sure all PR Checks have passed.
2. Pull down all related branches.
3. Navigate to `tests--setup-jsdom-testing`
4. Open the test file in `src/js/index.spec.js`
5. Run `npm test`

FlashGooden committed:
refactor: removed grid placeholders

FlashGooden committed:
feat: adding click Handling abilities
- we create a function called "handleClickCell" , this will be called once our game cell is clicked on. It has a parameter called CELL that will  be the node we will be updating. the cell new value will be "X" if clicked on
- starting from line 8 we will be gathering all game cells using our query selector, each of these cells will be given a clickhandler that will call on the "handleClickCell" function

FlashGooden committed:
style: clean up console logs

FlashGooden committed:
test: added skeleton test file for index.js

mcksheridan committed:
Chore: Install JSDOM

FSA2021-68

mcksheridan committed:
Test: Add example test with JSDom

Imported JSDom into the test file and added a simple test for the content of
a 'Hello, world' paragraph

FSA2021-68

mcksheridan committed:
Style: Add space after the console debug message

FSA2021-68

mcksheridan committed:
Test: Prepare for DOM testing with index.html

Exported the handleClickCell() function and imported it to the testing file.
Imported the file system API and used it to store the index.html file as a string,
which I then passed into a new JSDOM object for use in future tests.

FSA2021-68

mcksheridan committed:
Test: Add test for empty grid cell

FSA2021-68

mcksheridan committed:
Test: Add test for marking cell with X

FSA2021-68

mcksheridan committed:
Test: Add test for marking all grid cells with Xs

FSA2021-68

mcksheridan committed:
Chore: Changes to package lock

jordanleven commented on src/js/index.spec.js:
```diff
@@ -0,0 +1,42 @@
+
+const { JSDOM } = jsdom;
+
+const indexHtmlFile = fs.readFileSync('./src/html/index.html').toString();
```
This is a nifty approach, but you can also make your own HTML with the grid as part of this test. The issue with importing your HTML into this test is that now your tests will fail if someone modified your HTML file (something that you might want to catch as part of [an integration](https://www.cypress.io) test instead of a unit test).

jordanleven commented on src/js/index.spec.js:
```diff
@@ -0,0 +1,42 @@
+const dom = new JSDOM(indexHtmlFile);
+
+describe('index.js tests will go here', () => {
+  it('expects first test to equal 1', () => {
```
One suggestion for placeholder tests: use the Jest [`todo`](https://jestjs.io/docs/en/api#testtodoname) method instead of a regular test. Otherwise, folks might run tests and forget that this one is just a placeholder (since it appears as a regular test) when running them in the command line.

FlashGooden replied:
Copy, Thanks Jordan. I wrote this just to get something setup. Noted

jordanleven commented on src/js/index.spec.js:
```diff
@@ -0,0 +1,42 @@
+describe('replacing text in the paragraph tag', () => {
+  const testDom = new JSDOM('<doctype html><p>Hello, world</p>');
+  const testDomParagraph = testDom.window.document.querySelector('p');
+  it('expects the paragraph to have a value of "Hello, world"', () => {
```
I wasn't quick catching what this test is checking. What do you think?

mcksheridan replied:
This is just a test I made to get a feel for using jsdom. I'm pretty sure I shouldn't have committed it in the first place. 😅

jordanleven commented on src/js/index.spec.js:
```diff
@@ -0,0 +1,42 @@
+  // As there are multiple grid cell elements, this will only select the first one.
+  const gridCell = dom.window.document.querySelector('.grid-cell');
+  const allGameCells = dom.window.document.querySelectorAll('.grid-cell');
+  it('Shows an empty grid cell before any choices have been made', () => {
```
:chefskiss: This is perfect! Always make assertions to protect against incorrectly-passing tests. Nicely done!

jordanleven commented on src/js/index.spec.js:
```diff
@@ -0,0 +1,42 @@
+    expect(gridCell.textContent).toBe('');
+  });
+  it('Marks a grid cell with an "x"', () => {
+    handleClickCell(gridCell);
```
❤️ This is great! Out of curiosity, what would happen if you replaced line 34 with something like this:

```
gridCell.click()
```

jordanleven reviewed (commented):
Hey @mcksheridan - looking great! This is a really nice approach to checking the interactivity of your script. And the organization of your commits is 💯 .

I left a few comments throughout your PR. Let me know what you think!

mcksheridan committed:
Test: Replace HTML import

As the HTML for our grid is likely to evolve, directly importing the HTML file means that
the HTML our tests are based off of will continuously change. This could cause tests that
pass with consideration to our current code, to fail in the future. To prevent this from
happening, I removed the HTML import that used the file system and replaced it with a
gridGame variable written directly into the describe block for handling the click cell
function.

FSA2021-68

mcksheridan committed:
Test: Replace placeholder test with todo block

FSA2021-68

mcksheridan committed:
Test: Remove placeholder test

FSA2021-68

mcksheridan committed:
Test: Remove placeholder jsdom test

FSA2021-68

mcksheridan committed:
Test: Move the click event listener for allGameCells

Moved the event listener for allGameCells to its parent function (i.e. the describe block)
for use by all tests in the handle click cell function describe block.

FSA2021-68

mcksheridan committed:
Test: Add test for making one cell with an X by clicking

Separated the test for marking one grid cell with an X into two tests:
the original test (i.e. calling handleClickCell directly for one cell)
the new test (calling click on a gridCell)

FSA2021-68

FlashGooden commented:
Merged this in with the parent branch already. Rebased and merged into main
